runtime/gc_modron_startup/mminit.cpp

```diff
@@ -1101,54 +1101,59 @@ gcInitializeXmxXmdxVerification(J9JavaVM *javaVM, IDATA* memoryParameters, bool
+			} else {
+				extensions->heapCeiling = LOW_MEMORY_HEAP_CEILING;
+			}
 		} else {
-			extensions->heapCeiling = LOW_MEMORY_HEAP_CEILING;
+			extensions->heapCeiling = NON_SCALING_LOW_MEMORY_HEAP_CEILING;
 		}
-	} else {
-		extensions->heapCeiling = NON_SCALING_LOW_MEMORY_HEAP_CEILING;
-	}
 
 #if defined(J9ZOS39064)
-	{
-		/*
-		 * In order to support Compressed References ZOS should support one of:
-		 * - 2_TO_64 to support heaps allocation below 64GB
-		 * - 2_TO_32 to support heaps allocation below 32GB
-		 */
-		UDATA maxHeapForCR = 0;
-		switch (getUserExtendedPrivateAreaMemoryType()) {
-		case ZOS64_VMEM_ABOVE_BAR_GENERAL:
-		default:
-			/* options are not supported, heap allocation will fail eventually */
-			break;
-		case ZOS64_VMEM_2_TO_32G:
-			maxHeapForCR = DEFAULT_LOW_MEMORY_HEAP_CEILING;
-			break;
-		case ZOS64_VMEM_2_TO_64G:
-			maxHeapForCR = LOW_MEMORY_HEAP_CEILING;
-			break;
-		}
-
-		if (0 == maxHeapForCR) {
-			/* Redirector should not allow to run Compressed References JVM if options are not available */
-			Assert_MM_unreachable();
-			/* Fail to initialize if assertions are off */
-			return JNI_ERR;
-		}
+		{
+			/*
+			 * In order to support Compressed References ZOS should support one of:
+			 * - 2_TO_64 to support heaps allocation below 64GB
+			 * - 2_TO_32 to support heaps allocation below 32GB
+			 */
+			UDATA maxHeapForCR = 0;
+			switch (getUserExtendedPrivateAreaMemoryType()) {
+			case ZOS64_VMEM_ABOVE_BAR_GENERAL:
+			default:
+				/* options are not supported, heap allocation will fail eventually */
+				break;
+			case ZOS64_VMEM_2_TO_32G:
+				maxHeapForCR = DEFAULT_LOW_MEMORY_HEAP_CEILING;
+				break;
+			case ZOS64_VMEM_2_TO_64G:
+				maxHeapForCR = LOW_MEMORY_HEAP_CEILING;
+				break;
+			}
 
-		/* Adjust heap ceiling value if it is necessary */
-		if (extensions->heapCeiling > maxHeapForCR) {
-			extensions->heapCeiling = maxHeapForCR;
+			if (0 == maxHeapForCR) {
+				/* Redirector should not allow to run Compressed References JVM if options are not available */
+				Assert_MM_unreachable();
+				/* Fail to initialize if assertions are off */
+				return JNI_ERR;
+			}
+	
+			/* Adjust heap ceiling value if it is necessary */
+			if (extensions->heapCeiling > maxHeapForCR) {
+				extensions->heapCeiling = maxHeapForCR;
+			}
 		}
-	}
 #endif /* defined(J9ZOS39064) */
 
-	if (extensions->memoryMax > (extensions->heapCeiling - J9GC_COMPRESSED_POINTER_NULL_REGION_SIZE)) {
-		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTION_OVERFLOW, displayXmxOrMaxRAMPercentage(memoryParameters));
-		return JNI_ERR;
+		if (extensions->memoryMax > (extensions->heapCeiling - J9GC_COMPRESSED_POINTER_NULL_REGION_SIZE)) {
+			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTION_OVERFLOW, displayXmxOrMaxRAMPercentage(memoryParameters));
+			return JNI_ERR;
+		}
+	} else {
+		/* Constructor initialized heapCeiling to a compressed refs default - reset it if not compressed */
+		extensions->heapCeiling = 0;
```

Comment:
@charliegracie Can you comment on the approach here.  I'm not clear on whether there's a different (better?) way to do this.

Reply:
Spoke with @charliegracie - I'll move the reset into the initializer.

Reply:
This will unfortunately require the OMR change to happen first - marking WIP until promotion.

Reply:
Actually doesn't require the OMR portion since we're not compiling mixed mode anywhere. Ready to merge.